treeherder/auth/backends.py

```diff
@@ -99,32 +95,40 @@ def authenticate(self, auth_header=None, host=None, port=None):
 
         if result["status"] != "auth-success":
             logger.warning("Error logging in: {}".format(result["message"]))
-            raise TaskclusterAuthenticationFailed(result["message"])
+            raise TaskclusterAuthenticationFailedException(result["message"])
 
+        client_id = result["clientId"]
+        email = self._get_email(client_id)
         # TODO: remove this size limit when we upgrade to django 1.10
         # in Bug 1311967
-        email = self._get_email(result)
-        if len(email) <= 30:
-            username = email
+        if len(client_id) <= 30:
+            username = client_id
         else:
             username = base64.urlsafe_b64encode(
-                hashlib.sha1(smart_bytes(email)).digest()
+                hashlib.sha1(smart_bytes(client_id)).digest()
                 ).rstrip(b'=')[-30:]
 
         try:
-            # Find the user by their email.
-            user = self._get_user(email)
-            user.username = username
+            if scope_match(result["scopes"],
+                           [["assume:mozilla-user:{}".format(email)]]):
+                # Find the user by their email.
+                user = self._get_user(email)
```

Comment:
``self._get_user()`` actually raises on ``if not users:`` so I think that should hit this OK.

The test called ``test_create_missing_user`` tests this.  But I also created a test explicitly for the raising in ``_get_user()``